File: MvpCore/Events/EventAggregator.cs
using System.Collections.Concurrent;

namespace MvpCore.Events
{
    public class EventAggregator : IEventPublisher
    {
        private readonly ConcurrentDictionary<Type, List<Subscription>> _subscriptions = new();

        public void Publish<TEvent>(TEvent @event)
        {
            var eventType = typeof(TEvent);
            var allTypes = GetBaseTypesIncludingSelf(eventType);

            foreach (var type in allTypes)
            {
                if (_subscriptions.TryGetValue(type, out var handlers))
                {
                    foreach (var handler in handlers.ToArray())
                    {
                        handler.Invoke(@event!);
                    }
                }
            }
        }
        public IDisposable Subscribe<TEvent>(Action<TEvent> handler)
        {
            var type = typeof(TEvent);
            var sub = new Subscription<TEvent>(handler, RemoveSubscription);
            _subscriptions.AddOrUpdate(
                type,
                _ => new List<Subscription> { sub },
                (_, existing) =>
                {
                    existing.Add(sub);
                    return existing;
                });
            return sub;
        }

        private void RemoveSubscription(Subscription subscription)
        {
            if (_subscriptions.TryGetValue(subscription.EventType, out var subs))
            {
                _subscriptions[subscription.EventType] = subs.Where(s => s != subscription).ToList();
            }
        }

        private static IEnumerable<Type> GetBaseTypesIncludingSelf(Type type)
        {
            var types = new HashSet<Type> { type };
            var current = type.BaseType;

            while (current != null)
            {
                types.Add(current);
                current = current.BaseType;
            }

            foreach (var interfaceType in type.GetInterfaces())
            {
                types.Add(interfaceType);
            }

            return types;
        }

        private abstract class Subscription : IDisposable
        {
            public Type EventType { get; }
            protected Subscription(Type eventType) => EventType = eventType;
            public abstract void Invoke(object @event);
            public abstract bool Matches(object handler);
            public abstract void Dispose();
        }

        private sealed class Subscription<TEvent> : Subscription
        {
            private Action<TEvent>? _handler;
            private readonly Action<Subscription> _removeAction;

            public Subscription(Action<TEvent> handler, Action<Subscription> removeAction)
                : base(typeof(TEvent))
            {
                _handler = handler;
                _removeAction = removeAction;
            }

            public override void Invoke(object @event)
            {
                if (@event is TEvent typedEvent)
                {
                    _handler?.Invoke(typedEvent);
                }
            }

            public override bool Matches(object handler) =>
                ReferenceEquals(_handler, handler as Action<TEvent>);

            public override void Dispose()
            {
                _removeAction(this);
                _handler = null;
            }
        }
    }
}

File: mvp-core/src/events/EventAggregator.ts
import { IEventPublisher } from "../interfaces/IEventPublisher";
import { eventHierarchy } from "./EventTypes";

export class EventAggregator implements IEventPublisher {
    private subscriptions = new Map<string, Set<Function>>();

    publish<T extends object>(event: T): void {
        console.log("EventAggregator publish");

        const eventTypes = this.getEventTypeHierarchy(event);
        eventTypes.forEach(type => {
            const handlers = this.subscriptions.get(type) || new Set();
            handlers.forEach(handler => handler(event));
        });
    }

    subscribe<T extends object>(
        eventType: new (...args: any[]) => T,
        handler: (event: T) => void
    ): () => void {
        console.log("EventAggregator subscribe");

        const typeName = eventType.name;
        const handlers = this.subscriptions.get(typeName) || new Set<Function>();
        handlers.add(handler);
        this.subscriptions.set(typeName, handlers);
        return () => this.unsubscribe(typeName, handler);
    }

    private unsubscribe(typeName: string, handler: Function): void {
        console.log("EventAggregator unsubscribe");

        const handlers = this.subscriptions.get(typeName);
        if (handlers) {
            handlers.delete(handler);
            if (handlers.size === 0) {
                this.subscriptions.delete(typeName);
            }
        }
    }

    private getEventTypeHierarchy(event: object): string[] {
        console.log("EventAggregator getEventTypeHierarchy");

        const types: string[] = [];
        const eventName = event.constructor.name;
        console.log("eventName: " + eventName);

        // Add concrete type
        types.push(eventName);

        // Add registered base types
        if (eventHierarchy[eventName]) {
            types.push(...eventHierarchy[eventName]);
        }

        return types;
    }
}